ARTEMIS-2308 Support exporting metrics

jbertram opened the pull request:
Based on the conversation from #2626 I decided to pursue Micrometer integration which required a completely different approach than the one I took before. There is now no direct integration with Prometheus. The broker uses Micrometer APIs and a plugin structure was added to support specific integrations. A Prometheus plugin is available at https://github.com/jbertram/artemis-prometheus-metrics-plugin.

franz1981 commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/server/metrics/MetricsManager.java:
```diff
@@ -0,0 +1,125 @@
+import org.apache.activemq.artemis.api.core.management.ResourceNames;
+import org.apache.activemq.artemis.core.server.ActiveMQServerLogger;
+
+public class MetricsManager {
```
That's a huge comment; API change proposal...
```java
   @FunctionalInterface
   public interface MetricGaugeBuilder {
      void register(String metricName, Object state, ToDoubleFunction f, String description);
   }

   public void registerQueueGauge(String address, String queue, Consumer<MetricGaugeBuilder> builder) {
      final List<Meter> newMeters = new ArrayList<>();
      builder.accept((metricName, state, f, description) -> {
         Meter meter = Gauge
            .builder("artemis." + metricName, state, f)
            .tag("broker", brokerName)
            .tag("address", address)
            .tag("queue", queue)
            .description(description)
            .register(meterRegistry);
         newMeters.add(meter);
      });
      final String resource = ResourceNames.QUEUE + queue;
      this.meters.compute(resource, (s, meters) -> {
         if (meters == null) {
            meters = new ArrayList<>();
         }
         meters.addAll(newMeters);
         return meters;
      });
   }

   public void registerAddressGauge(String address, Consumer<MetricGaugeBuilder> builder) {
      final List<Meter> newMeters = new ArrayList<>();
      builder.accept((metricName, state, f, description) -> {
         Meter meter = Gauge
            .builder("artemis." + metricName, state, f)
            .tag("broker", brokerName)
            .tag("address", address)
            .description(description)
            .register(meterRegistry);
         newMeters.add(meter);
      });
      final String resource = ResourceNames.ADDRESS + address;
      this.meters.compute(resource, (s, meters) -> {
         if (meters == null) {
            meters = new ArrayList<>();
         }
         meters.addAll(newMeters);
         return meters;
      });
   }

   public void registerBrokerGauge(Consumer<MetricGaugeBuilder> builder) {
      final List<Meter> newMeters = new ArrayList<>();
      builder.accept((metricName, state, f, description) -> {
         Meter meter = Gauge
            .builder("artemis." + metricName, state, f)
            .tag("broker", brokerName)
            .description(description)
            .register(meterRegistry);
         newMeters.add(meter);
      });
      final String resource = ResourceNames.BROKER + "." + brokerName;
      this.meters.compute(resource, (s, meters) -> {
         if (meters == null) {
            meters = new ArrayList<>();
         }
         meters.addAll(newMeters);
         return meters;
      });
   }
```

That could be used with:
```java
if (server != null && server.getMetricsManager() != null) {
         server.getMetricsManager().registerQueueGauge(queueName, addressName, builder -> {
            builder.register("message.count", pendingMetrics, metrics -> Double.valueOf(pendingMetrics.getMessageCount()), QueueControl.MESSAGE_COUNT_DESCRIPTION);
            builder.register(//...other useful metrics
```

jbertram replied:
Maybe it's just me, but I think the existing implementation is easier to read.  :)

Is there a particular benefit to your implementation? I'm assuming it has better performance for some esoteric reason. :)

franz1981 replied:
Maybe :P It has a better behaviour under load, because it won't call many times put on the concurrent hash map, just once...

franz1981 replied:
And it is more safe concurrently, given that the overall configuration is consider an atomic operation

jbertram commented on artemis-core-client/src/main/java/org/apache/activemq/artemis/api/core/management/AddressControl.java:
```diff
@@ -101,19 +103,19 @@
    String[] getBindingNames() throws Exception;
 
-   @Attribute(desc = "number of messages added to all the queues for this address")
+   @Attribute(desc = "number of messages currently in all queues bound to this address (includes scheduled, paged, and in-delivery messages)")
```
I didn't refactor the description here because I'm not actually exported this metric since it will be easy to aggregate all the message counts of all the queues on the address in whatever system the metrics are exported to.

jbertram commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java:
```diff
@@ -3852,6 +3859,41 @@ void scheduleSlowConsumerReaper(AddressSettings settings) {
+
+      if (server != null && server.getMetricsManager() != null) {
+         MetricsManager metricsManager = server.getMetricsManager();
+         metricsManager.registerQueueGauge("message.count", addressName, queueName, pendingMetrics, metrics -> Double.valueOf(pendingMetrics.getMessageCount()), QueueControl.MESSAGE_COUNT_DESCRIPTION);
```
I'm not entirely sure what you mean. Do you mean declaring something like:

`static final String MESSAGE_COUNT = "message.count";`

And then using `MESSAGE_COUNT` in the method invocation?

jbertram commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/server/metrics/MetricsManager.java:
```diff
@@ -0,0 +1,125 @@
+
+   private MeterRegistry meterRegistry;
+
+   private Map<String, List<Meter>> meters = new HashMap<>();
```
Agreed. I'll change that.

jbertram commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/server/metrics/MetricsManager.java:
```diff
@@ -0,0 +1,125 @@
+      return meterRegistry;
+   }
+
+   public void registerQueueGauge(String metricName, String address, String queue, Object object, ToDoubleFunction f, String description) {
```
This isn't possible because a `QueuePendingMessageMetrics` isn't always passed into the method (e.g. for all the queue metrics that aren't tracked by the `QueuePendingMessageMetrics` object).

michaelandrepearce commented on docs/user-manual/en/metrics.md:
```diff
@@ -0,0 +1,104 @@
+Important runtime metrics have been instrumented via the Micrometer API, and
+all a user needs to do is implement `org.apache.activemq.artemis.core.server.metrics.ActiveMQMetricsPlugin`
+in order to instantiate and configure a `io.micrometer.core.instrument.MeterRegistry`
+implementation. Many implementations of `MeterRegistr` are available from
```
sentance seems to end abrubtly i assume you meant to put a link to the micrometer web page with a list of these?

michaelandrepearce commented:
Great stuff. +1 on this from me. Nice abstraction from implementations.

michaelandrepearce commented on artemis-server/src/main/resources/schema/artemis-configuration.xsd:
```diff
@@ -1012,6 +1012,33 @@
             </xsd:complexType>
          </xsd:element>
 
+         <xsd:element name="metrics-plugin" maxOccurs="1" minOccurs="0">
```
Is the test version of artemis-configuration.xsd updated also, to keep them in sync.

clebertsuconic commented:
It would be nice now to have the components repo, and an option to include it with the installation.
Great stuff...


is this ready to merge?

franz1981 commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/AddressInfo.java:
```diff
@@ -191,4 +194,16 @@ public long getUnRoutedMessageCount() {
+   }
+
+   public void unregisterMeters(MetricsManager metricsManager) {
+      if (metricsManager != null) {
```
Given that `metricsManager` is volatile, it should be called in a local variable and if not null, used to call its method

franz1981 replied:
Ups sorry, reading it from the web I haven't noticed from where `metricsManager` was coming

franz1981 commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/server/metrics/MetricsManager.java:
```diff
@@ -0,0 +1,125 @@
+         .description(description)
+         .register(meterRegistry);
+
+      meters.add(meter);
```
This risk to be called by different threads: take a look to the implementation I've posted above (that is using the compute method, that is thread safe for a specific key).

jbertram commented:
I just pushed the latest changes:

- Fixed documentation (Mike)
- Fixed XSD (Mike)
- Addressed thread-safety issues (Franz)
- Added static finals for metric names (Franz)

clebertsuconic commented:
@jbertram there still work to be improved, mainly when we get the repository, but this is good as is for now. We can send more commits later.

michaelandrepearce commented on docs/user-manual/en/metrics.md:
```diff
@@ -0,0 +1,105 @@
+**Address**
+
+- routed.message.count
+- unrouted.message.count
```
@jbertram what about paging size?
